homeassistant/util/unit_system.py

```diff
@@ -116,6 +141,39 @@ def as_dict(self) -> dict:
             VOLUME: self.volume_unit
         }
 
+    def convert(self, value, unit):
+        """Generic conversion method."""
+        converted = None
+        if (unit in (TEMP_CELSIUS, TEMP_FAHRENHEIT) and
+                unit != self.temperature_unit):
+            # Convert temperature if we detect one
+            prec = len(value) - value.index('.') - 1 if '.' in value else 0
+            try:
+                temp = self.temperature(float(value), unit)
+            except ValueError:
+                return None
+            value = str(round(temp) if prec == 0 else round(temp, prec))
```

Comment:
Value should not be converted to a string in here, this should remain a number. It should be converted to a string inside the entity class.

Reply:
This is the old implementation moved from the entity to the unit system. The only modifications I did to this code were to move the try/except as you requested in an earlier pull.